runtime/src/main/java/org/corfudb/protocols/wireprotocol/ClientHandshakeHandler.java

```diff
@@ -0,0 +1,223 @@
+                return;
+            }
+
+            if (this.handshakeComplete.get()) {
+                // If handshake completed successfully, but still a message came through this handler,
+                // send on to the next handler in order to avoid message loss.
+                super.channelRead(ctx, m);
+                return;
+            }
+
+            CorfuPayloadMsg<HandshakeResponse> handshakeResponse;
+
+            try {
+               handshakeResponse = (CorfuPayloadMsg<HandshakeResponse>) m;
+            } catch (ClassCastException e) {
+                log.warn("Non-handshake message received by handshake handler. Send upstream only " +
+                        "if handshake succeeded.");
+                if (this.handshakeComplete.get()) {
+                    // Only send upstream if handshake is complete.
+                    super.channelRead(ctx, m);
+                }
+                return;
+            }
+
+            UUID serverId = handshakeResponse.getPayload().getServerId();
+            String corfuVersion = handshakeResponse.getPayload().getCorfuVersion();
+
+            // Handshake Validation
+            if (!this.nodeId.equals(serverId)) {
+                // Validation failed, client opened a socket to server with id
+                // 'nodeId', instead server's id is 'serverId'
+                log.error("Handshake validation failed. Server node id mismatch.");
+                log.debug("Client opened socket to server {" + this.nodeId
+                        + "} instead, connected to: {" + serverId + "}");
+                this.fireHandshakeFailed(ctx);
+                return;
+            }
+
+            log.info("Handshake succeeded. Server Corfu Version: {" + corfuVersion + "}");
+            // Flush messages in queue
+            log.debug("There are {" + this.messages.size() + "} messages in queue to " +
+                    "be flushed.");
+            for (CorfuMsg message : this.messages) {
+                ctx.writeAndFlush(message);
+            }
+
+            // Remove this handler from the pipeline; handshake is completed.
+            log.info("Removing handshake handler from pipeline.");
+            ctx.pipeline().remove(this);
+            this.fireHandshakeSucceeded();
+        }
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx)
+            throws Exception {
+        log.info("Outgoing connection established to: " + ctx.channel().remoteAddress());
+
+        // Write the handshake & add a timeout listener.
+        CorfuMsg handshake = CorfuMsgType.HANDSHAKE_INITIATE
+                .payloadMsg(new HandshakeMsg(this.clientId, this.nodeId));
+
+        log.info("Initiate handshake. Send handshake message.");
+        ChannelFuture f = ctx.writeAndFlush(handshake);
+        f.addListener(new ChannelFutureListener() {
+            @Override
+            public void operationComplete(ChannelFuture future)
```

Comment:
Same re: ReadTimeoutHandler on the server side.

Reply:
Done